src/pymor/algorithms/samdp.py

```diff
@@ -0,0 +1,448 @@
+    y
+        Refined left eigenvector of matrix pair (A, E).
+    theta
+        Refined eigenvalue of matrix pair (A, E).
+    residual
+        Residual of the computed triplet.
+    """
+
+    i = 0
+    nrq = 1
+    while nrq > tol and i < maxiter:
+        i = i + 1
+        Ex = E.apply(x)
+        Ey = E.apply_adjoint(y)
+        tEmA = theta * E - A
+        x_rqi = tEmA.apply_inverse(Ex)
+        v_rqi = tEmA.apply_inverse_adjoint(Ey)
+
+        x_rqi.scal(1 / x_rqi.norm())
+        v_rqi.scal(1 / v_rqi.norm())
+
+        Ax_rqi = A.apply(x_rqi)
+        Ex_rqi = E.apply(x_rqi)
+
+        x_rq = (v_rqi.dot(Ax_rqi) / v_rqi.dot(Ex_rqi))[0][0]
+        if not np.isfinite(x_rq):
+            x_rqi = x
+            v_rqi = y
+            x_rq = theta + 1e-10
+
+        rqi_res = Ax_rqi - Ex_rqi * x_rq
+        if np.abs(np.imag(x_rq)) / np.abs(x_rq) < imagtol:
+            rx_rqi = np.real(x_rqi)
+            rx_rqi.scal(1 / rx_rqi.norm())
+
+            rres = A.apply(rx_rqi) - E.apply(rx_rqi) * np.real(x_rq)
+            nrr = rres.norm()
+            if nrr < rqi_res.norm():
+                x_rqi = rx_rqi
+                v_rqi = np.real(v_rqi)
+                v_rqi.scal(1 / v_rqi.norm())
+                x_rq = np.real(x_rq)
+                rqi_res = rres
+
+        x = x_rqi
+        y = v_rqi
+        theta = x_rq
+        nrq = rqi_res.norm()
+
+        if not np.isfinite(nrq):
+            nrq = 1
+    if nrq < init_res:
+        return x_rqi, v_rqi, x_rq, nrq
+    else:
+        return x, y, theta, init_res
+
+
+def select_max_eig(H, G, X, V, B, C, which):
+    """Compute poles sorted from largest to smallest residual.
+
+    Parameters
+    ----------
+    H
+        The |Numpy array| H from the samdp algorithm.
+    G
+        The |Numpy array| G from the samdp algorithm.
+    X
+        A |VectorArray| describing the orthogonal search space used in the samdp algorithm.
+    V
+        A |VectorArray| describing the orthogonal search space used in the samdp algorithm.
+    B
+        The |VectorArray| B from the corresponding LTI system modified by deflation.
+    C
+        The |VectorArray| C from the corresponding LTI system modified by deflation.
+    which
+        A string that indicates which poles to select. See :func:`samdp`.
+
+    Returns
+    -------
+    poles
+        A |NumPy array| containing poles sorted according to the chosen strategy.
+    rightevs
+        A |NumPy array| containing the right eigenvectors of the computed poles.
+    leftevs
+        A |NumPy array| containing the left eigenvectors of the computed poles.
+    """
+
+    D, Vs = spla.eig(H, G)
+    idx = np.argsort(D)
+    DP = D[idx]
+    Vs = Vs[:, idx]
+
+    Dt, Vt = spla.eig(H.conj().T, G.conj().T)
+    Vt = Vt[:, np.argsort(Dt)]
+
+    X = X.lincomb(Vs.T)
+    V = V.lincomb(Vt.T)
+
+    V.scal(1 / V.norm())
+    X.scal(1 / X.norm())
+    residue = spla.norm(C.dot(X), 2, 0) * spla.norm(V.dot(B), 2, 1)
```

Comment:
The 2-norm is the default for vectors.
```suggestion
    residue = spla.norm(C.dot(X), axis=0) * spla.norm(V.dot(B), axis=1)
```